into_make_service_with_connect_info not compile anymore

<!--
Thank you for reporting an issue.

Please fill in as much of the template below as you're able.
-->

after this PR https://github.com/tokio-rs/axum/pull/2941

with `tap_io` called on `listener`, 

`axum::serve(listener, app.into_make_service_with_connect_info::<SocketAddr>()).await.unwrap();`   not compile anymore
 
if comment out the tap_io call, it will pass the compilation

```rust
        /*.tap_io(|tcp_stream| {
            if let Err(err) = tcp_stream.set_nodelay(true) {
                eprintln!("failed to set TCP_NODELAY on incoming connection: {err:#}");
            }
        })*/;
```

- [x] I have looked for existing issues (including closed) about this

## Bug Report

### Version

main branch

### Platform

linux

### Crates

<!--
If known, please specify the affected axum crates. Otherwise, delete this
section.
-->

### Description

I tried this code:

```rust
use axum::{response::Html, routing::get, Router};
use axum::extract::connect_info::ConnectInfo;
use axum::serve::ListenerExt;
use std::net::SocketAddr;

#[tokio::main]
async fn main() {
    // build our application with a route
    let app = Router::new().route("/", get(handler));

    // run it
    let listener = tokio::net::TcpListener::bind("127.0.0.1:3000")
        .await
        .unwrap();
    println!("listening on {}", listener.local_addr().unwrap());

    let listener = listener
        .tap_io(|tcp_stream| {
            if let Err(err) = tcp_stream.set_nodelay(true) {
                eprintln!("failed to set TCP_NODELAY on incoming connection: {err:#}");
            }
        });

    axum::serve(listener, app.into_make_service_with_connect_info::<SocketAddr>()).await.unwrap();
}

async fn handler(ConnectInfo(addr): ConnectInfo<SocketAddr>) -> Html<String> {
    Html(format!("<h1>Hello, World! {}</h1>", addr))
}
```

cargo

```toml
[package]
name = "axum-connection-info"
version = "0.1.0"
edition = "2021"

[dependencies]
axum = { git = "https://github.com/tokio-rs/axum", branch = "main" }
tokio = { version = "1.41.1", features = ["full"] }
```

---------------------------------------------------

92 | pub fn serve<L, M, S>(listener: L, make_service: M) -> Serve<L, M, S>
   |        ----- required by a bound in this function
...
95 |     M: for<'a> Service<IncomingStream<'a, L>, Error = Infallible, Response = S>,
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `serve`

error[E0277]: `Serve<TapIo<TcpListener, {closure@main.rs:18:17}>, IntoMakeServiceWithConnectInfo<..., ...>, ...>` is not a future
  --> src/main.rs:24:84
   |
24 |     axum::serve(listener, app.into_make_service_with_connect_info::<SocketAddr>()).await.unwrap();
   |     -------------------------------------------------------------------------------^^^^^
   |     |                                                                             ||
   |     |                                                                             |`Serve<TapIo<TcpListener, {closure@main.rs:18:17}>, IntoMakeServiceWithConnectInfo<..., ...>, ...>` is not a future
   |     |                                                                             help: remove the `.await`
   |     this call returns `Serve<TapIo<tokio::net::TcpListener, {closure@src/main.rs:18:17: 18:29}>, IntoMakeServiceWithConnectInfo<Router, std::net::SocketAddr>, AddExtension<Router, ConnectInfo<std::net::SocketAddr>>>`

--- a/axum/src/extract/connect_info.rs
+++ b/axum/src/extract/connect_info.rs
@@ -92,6 +92,17 @@ const _: () = {
             *stream.remote_addr()
         }
     }
+
+    impl<'a, L, F> Connected<serve::IncomingStream<'a, serve::TapIo<L, F>>> for L::Addr
+    where
+        L: serve::Listener,
+        L::Addr: Clone + Sync + 'static,
+        F: FnMut(&mut L::Io) + Send + 'static,
+    {
+        fn connect_info(stream: serve::IncomingStream<'a, serve::TapIo<L, F>>) -> Self {
+            stream.remote_addr().clone()
+        }
+    }
 };
 
 impl Connected<SocketAddr> for SocketAddr {

--- a/axum/src/serve.rs
+++ b/axum/src/serve.rs
@@ -434,6 +434,7 @@ mod tests {
         extract::connect_info::Connected,
         handler::{Handler, HandlerWithoutStateExt},
         routing::get,
+        serve::ListenerExt,
         Router,
     };
 
@@ -452,12 +453,27 @@ mod tests {
 
         let addr = "0.0.0.0:0";
 
+        let tcp_nodelay_listener = || async {
+            TcpListener::bind(addr).await.unwrap().tap_io(|tcp_stream| {
+                if let Err(err) = tcp_stream.set_nodelay(true) {
+                    eprintln!("failed to set TCP_NODELAY on incoming connection: {err:#}");
+                }
+            })
+        };
+
         // router
         serve(TcpListener::bind(addr).await.unwrap(), router.clone());
+        serve(tcp_nodelay_listener().await, router.clone())
+            .await
+            .unwrap();
         serve(UnixListener::bind("").unwrap(), router.clone());
 
         serve(
             TcpListener::bind(addr).await.unwrap(),
+            router.clone().into_make_service(),
+        );
+        serve(
+            tcp_nodelay_listener().await,
             router.clone().into_make_service(),
         );
         serve(
@@ -472,16 +488,27 @@ mod tests {
                 .into_make_service_with_connect_info::<std::net::SocketAddr>(),
         );
         serve(
+            tcp_nodelay_listener().await,
+            router
+                .clone()
+                .into_make_service_with_connect_info::<std::net::SocketAddr>(),
+        );
+        serve(
             UnixListener::bind("").unwrap(),
             router.into_make_service_with_connect_info::<UdsConnectInfo>(),
         );
 
         // method router
         serve(TcpListener::bind(addr).await.unwrap(), get(handler));
+        serve(tcp_nodelay_listener().await, get(handler));
         serve(UnixListener::bind("").unwrap(), get(handler));
 
         serve(
             TcpListener::bind(addr).await.unwrap(),
+            get(handler).into_make_service(),
+        );
+        serve(
+            tcp_nodelay_listener().await,
             get(handler).into_make_service(),
         );
         serve(
@@ -494,6 +521,10 @@ mod tests {
             get(handler).into_make_service_with_connect_info::<std::net::SocketAddr>(),
         );
         serve(
+            tcp_nodelay_listener().await,
+            get(handler).into_make_service_with_connect_info::<std::net::SocketAddr>(),
+        );
+        serve(
             UnixListener::bind("").unwrap(),
             get(handler).into_make_service_with_connect_info::<UdsConnectInfo>(),
         );
@@ -503,22 +534,29 @@ mod tests {
             TcpListener::bind(addr).await.unwrap(),
             handler.into_service(),
         );
+        serve(tcp_nodelay_listener().await, handler.into_service());
         serve(UnixListener::bind("").unwrap(), handler.into_service());
 
         serve(
             TcpListener::bind(addr).await.unwrap(),
             handler.with_state(()),
         );
+        serve(tcp_nodelay_listener().await, handler.with_state(()));
         serve(UnixListener::bind("").unwrap(), handler.with_state(()));
 
         serve(
             TcpListener::bind(addr).await.unwrap(),
             handler.into_make_service(),
         );
+        serve(tcp_nodelay_listener().await, handler.into_make_service());
         serve(UnixListener::bind("").unwrap(), handler.into_make_service());
 
         serve(
             TcpListener::bind(addr).await.unwrap(),
+            handler.into_make_service_with_connect_info::<std::net::SocketAddr>(),
+        );
+        serve(
+            tcp_nodelay_listener().await,
             handler.into_make_service_with_connect_info::<std::net::SocketAddr>(),
         );
         serve(

--- a/axum/src/serve/listener.rs
+++ b/axum/src/serve/listener.rs
@@ -102,7 +102,7 @@ impl<L: Listener> ListenerExt for L {}
 /// Return type of [`ListenerExt::tap_io`].
 ///
 /// See that method for details.
-pub struct TapIo<L: Listener, F> {
+pub struct TapIo<L, F> {
     listener: L,
     tap_fn: F,
 }

